src/ucp/proto/proto_init.c

```diff
@@ -196,180 +115,176 @@ ucp_proto_perf_envelope_make(const ucs_linear_func_t *funcs, uint64_t funcs_num,
 {
-    ucp_context_h context            = params->super.worker->context;
-    ucp_proto_perf_node_t *perf_node = NULL;
+    ucp_context_h context                 = params->super.worker->context;
+    ucp_proto_perf_factors_t perf_factors = UCP_PROTO_PERF_FACTORS_INITIALIZER;
+    ucp_proto_perf_node_t *reg_perf_node;
     const uct_md_attr_v2_t *md_attr;
     ucp_md_index_t md_index;
+    ucs_status_t status;
     const char *md_name;
 
-    *memreg_time = UCS_LINEAR_FUNC_ZERO;
-
     if (reg_md_map == 0) {
-        goto out;
+        return UCS_OK;
     }
 
     if (context->rcache != NULL) {
-        perf_node = ucp_proto_perf_node_new_data("rcache lookup", "");
-
-        *memreg_time = ucs_linear_func_make(context->config.ext.rcache_overhead,
-                                            0);
-
-        ucp_proto_perf_node_add_data(perf_node, "lookup", *memreg_time);
-
-        goto out;
+        perf_factors[cpu_factor_id] =
+                ucs_linear_func_make(context->config.ext.rcache_overhead, 0);
+        return ucp_proto_perf_add_funcs(perf, range_start, range_end,
+                                        perf_factors, NULL, "rcache lookup",
+                                        "");
     }
 
-    perf_node = ucp_proto_perf_node_new_data("mem reg", "");
+    reg_perf_node = ucp_proto_perf_node_new_data("mem reg", "");
 
     /* Go over all memory domains */
     ucs_for_each_bit(md_index, reg_md_map) {
         md_attr = &context->tl_mds[md_index].attr;
         md_name = context->tl_mds[md_index].rsc.md_name;
-        ucs_linear_func_add_inplace(memreg_time, md_attr->reg_cost);
-        ucs_trace("md %s reg_cost: " UCP_PROTO_PERF_FUNC_FMT, md_name,
+        ucs_linear_func_add_inplace(&perf_factors[cpu_factor_id],
+                                    md_attr->reg_cost);
+        ucs_trace("md %s reg: " UCP_PROTO_PERF_FUNC_FMT, md_name,
                   UCP_PROTO_PERF_FUNC_ARG(&md_attr->reg_cost));
-
-        ucp_proto_perf_node_add_data(perf_node, md_name, md_attr->reg_cost);
+        ucp_proto_perf_node_add_data(reg_perf_node, md_name, md_attr->reg_cost);
     }
 
     if (!ucs_is_pow2(reg_md_map)) {
         /* Multiple memory domains */
-        ucp_proto_perf_node_add_data(perf_node, "total", *memreg_time);
+        ucp_proto_perf_node_add_data(reg_perf_node, "total",
+                                     perf_factors[cpu_factor_id]);
     }
 
-out:
-    *perf_node_p = perf_node;
+    status = ucp_proto_perf_add_funcs(perf, range_start, range_end,
+                                      perf_factors, reg_perf_node,
+                                      perf_node_name, "%u mds",
+                                      ucs_popcount(reg_md_map));
+    ucp_proto_perf_node_deref(&reg_perf_node);
+    return status;
+}
+
+static ucp_proto_perf_factor_id_t
+ucp_proto_buffer_copy_cpu_factor_id(int local) {
+    return (local) ? UCP_PROTO_PERF_FACTOR_LOCAL_CPU :
+                     UCP_PROTO_PERF_FACTOR_REMOTE_CPU;
```

Comment:
2,3 - ok, thanks

1 - I am OK with moving `ucp_proto_buffer_copy_cpu_factor_id` to `proto_perf.h` since it doesn't relies on other entities like memory ops and types but I think that it is better to leave `ucp_proto_buffer_copy_factor_id` on the current place. In that PR I tried to place perf structure common interfaces to `proto_perf.h` and place all the logic of defining perf for certain proto in `proto_init.c`, WDYT?